src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,274 @@
+//! The PROXY protocol header compatibility.
+//! See spec: <https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt>
+//!
+//! Note: if you are setting a custom acceptor, `proxy_protocol_enabled` must be called after this is set.
+//! Safest to use directly before serve when all options are configured.
+//!
+//! # Example
+//!
+//! ```rust,no_run
+//! use axum::{routing::get, Router};
+//! use std::net::SocketAddr;
+//!
+//! #[tokio::main]
+//! async fn main() {
+//!    let app = Router::new().route("/", get(|| async { "Hello, world!" }));
+//!
+//!    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+//!    println!("listening on {}", addr);
+//!    axum_server::bind(addr)
+//!        .proxy_protocol_enabled()
+//!        .serve(app.into_make_service())
+//!        .await
+//!        .unwrap();
+//! }
+//! ```
+use std::future::poll_fn;
+use std::future::Future;
+use std::io;
+use std::net::IpAddr;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+use std::time::Duration;
+
+use http::HeaderValue;
+use http::Request;
+use ppp::{v2, HeaderResult, PartialResult};
+use tokio::io::AsyncReadExt;
+use tokio::io::ReadBuf;
+use tower_service::Service;
+
+use crate::accept::Accept;
+use std::fmt;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+pub mod future;
+use self::future::ProxyProtocolAcceptorFuture;
+
+pub trait Peekable {
+    fn poll_peek(
+        &mut self,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<std::io::Result<usize>>;
+}
+
+/// Note: Currently only supports the V2 PROXY header format.
+/// See proxy header spec: <https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt>
+pub(crate) async fn read_proxy_header<I>(mut stream: I) -> Result<(I, Option<IpAddr>), io::Error>
+where
+    I: Peekable + AsyncRead + Unpin,
+{
+    // Maximum expected size for PROXY header is
+    // unlikely to be larger than 512 bytes unless additional fields are added.
+    const BUFFER_SIZE: usize = 512;
```

Comment:
Why? You should parse the first n bytes and determine if it's protocol v1, v2, or drop/error the connection.